File: src/audit.ts
/** Random selection among safe scenarios and audit-only drift inputs. */

import { type AuditSample, ValidationError } from "./models.js";
import { type DriftResult, detectDrift } from "./statistics/drift.js";
import { systemRandom } from "./statistics/random.js";

/** Draw before human review; persist the decision and inclusion probability. */
export function selectAudit(options: { judgedSafe: boolean; fraction: number; random?: () => number }): boolean {
  if (!(options.fraction > 0 && options.fraction <= 1)) throw new ValidationError("Audit fraction must be in (0, 1]");
  return options.judgedSafe && (options.random ?? systemRandom)() < options.fraction;
}

/** Monitor completed independent audit clusters under predeclared baselines. */
export function auditDrift(samples: readonly AuditSample[], options: { baselines: Record<string, number>; costUpperBound?: number | null; alpha?: number; expectedShift?: number }): DriftResult {
  if (new Set(samples.map((s) => s.functionId)).size > 1) throw new ValidationError("Monitor one function identity at a time");
  for (let i = 1; i < samples.length; i++) {
    if (Date.parse(samples[i].timestamp) < Date.parse(samples[i - 1].timestamp)) throw new ValidationError("Audit scenarios must arrive in chronological order");
  }
  return detectDrift(samples.map((s) => ({ scenarioId: s.scenarioId, cluster: s.cluster, judgedSafe: s.judgedSafe, auditSelected: s.selectedForAudit,
    auditProbability: s.inclusionProbability, humanReviewed: s.humanReviewed, attempted: { ...s.attempted }, cost: s.cost })), options);
}

File: src/signal_sdk/audit.py
"""Random selection among safe episodes and audit-only drift inputs."""

from random import SystemRandom
from typing import Any

from .models import AuditSample
from .statistics.drift import detect_drift


def select_audit(*, judged_safe: bool, fraction: float, rng: Any = None) -> bool:
    """Draw before human review; persist the decision and inclusion probability."""
    if not 0 < fraction <= 1:
        raise ValueError("Audit fraction must be in (0, 1]")
    return bool(judged_safe and (rng or SystemRandom()).random() < fraction)


def audit_drift(samples: tuple[AuditSample, ...], *, baselines: dict[str, float],
                cost_upper_bound: float | None = None, **settings: Any) -> dict:
    """Monitor completed independent audit clusters under predeclared baselines."""
    if len({s.function_id for s in samples}) > 1:
        raise ValueError("Monitor one function identity at a time")
    if list(samples) != sorted(samples, key=lambda s: s.timestamp):
        raise ValueError("Audit episodes must arrive in chronological order")
    rows = [{"episode_id": s.episode_id, "cluster": s.cluster, "judged_safe": s.judged_safe,
             "audit_selected": s.selected_for_audit, "audit_probability": s.inclusion_probability,
             "human_reviewed": s.human_reviewed, "attempted": dict(s.attempted), "cost": s.cost}
            for s in samples]
    return detect_drift(rows, baselines=baselines, cost_upper_bound=cost_upper_bound, **settings)
